projects/client-vscode/src/webview.ts

```diff
@@ -1,51 +1,76 @@
-import { ClientConnector, connectClient, applyApi, Client } from '@remixproject/plugin/connector'
-import { PluginClient } from '@remixproject/plugin/client'
-import { Message } from '../../utils/src/types/message'
-import { Api, ApiMap } from '../../utils/src/types/api'
-
+import {
+  ClientConnector,
+  connectClient,
+  applyApi,
+  Client,
+  PluginClient,
+  Message,
+  Api,
+  ApiMap,
+  isHandshake,
+  PluginOptions,
+  checkOrigin
+} from '@remixproject/plugin'
 
 declare const acquireVsCodeApi: any
-
 /**
  * This Webview connector
  */
 export class WebviewConnector implements ClientConnector {
-  postMessage: (message: any) => void
-  constructor() {
-    // Get the acquireVsCodeApi injected by vscode
-    if (acquireVsCodeApi) {
-      this.postMessage = acquireVsCodeApi().postMessage
-    }
+  source: { postMessage: (message: any, origin?: string) => void }
+  origin: string
+  isVscode: boolean
+
+  constructor(private options: PluginOptions<any>) {
+    this.isVscode = !!acquireVsCodeApi
+    this.source = this.isVscode ? acquireVsCodeApi() : window.parent
   }
 
+
   /** Send a message to the engine */
   send(message: Partial<Message>) {
-    this.postMessage(message)
+    if (this.isVscode) {
+      this.source.postMessage(message)
+    } else if (this.origin || isHandshake(message)) {
+      const origin = this.origin || '*'
+      this.source.postMessage(message, origin)
```

Comment:
don't need to use `window.parent.` anymore ? like for https://github.com/ethereum/remix-plugin/blob/v0.3.0/projects/client-iframe/src/connector.ts#L21

Reply:
It's the default source when parent s not vscode. : line 26
`this.source = this.isVscode ? acquireVsCodeApi() : window.parent`